qiskit/circuit/quantumcircuit.py

```diff
@@ -499,6 +517,19 @@ def __eq__(self, other) -> bool:
             other, copy_operations=False
         )
 
+    def __getstate__(self):
+        # Rust's CircuitData is not picklable, so we convert it to a list first.
+        state = self.__dict__.copy()
+        state["_data"] = list(self._data)
+        del state["_intern_context"]
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__ = state
+        self._intern_context = InternContext()
+        data = self._new_data(self._data)
+        self._data = data
```

Comment:
I'm worried that this is going to make `__deepcopy__` unnecessarily expensive for large circuits.  It's fine for a demo, but I feel like we could provide Python-space methods on the two objects that let them be properly pickled in terms of each other when they're happening in the context of a full circuit.  We'll likely also want to do something similar with `__deepcopy__` such that it _can_ share the `InternContext` with the copy.

Reply:
Yeah agree on prioritisation.  I was worried that they can't be pickled without reference to each other, but actually they definitely can in their current form, so it won't be as hard as I feared.

Reply:
I tried implementing `__{get,set}state__` for both `CircuitData` and `InternContext`, but the [implementations were a bit messy](https://github.com/kevinhartman/qiskit-terra/commit/fe6eaf7b1b2ecab054820676d936517b297f90f2), and I'm not convinced it's worth having them. The patch linked there is not part of this PR.

Instead, in f52300b I've overridden `__deepcopy__` to copy circuit instructions one at a time from the source to the destination using an iterator so we can avoid pulling everything into a `list` when we're not actually going to serialize anything.

As you'll also see in that patch, I tried reusing the same `InternContext` instance, but I ran into trouble with primitives jobs (which use thread pool executors for concurrency), so it's commented-out. Specifically, I was experiencing a race condition where `self.intern_context.as_ref(py)` was already borrowed, which presented non-deterministically when running the test suite. This seems odd, given that the GIL is always held when this is borrowed. The only thing I could think of is if somehow the thread is being preempted during the borrow, and another thread is grabbing the GIL and then borrowing. I don't know enough about Python's threading model to know if this could be happening. My assumption was that holding the GIL was somewhat like a cooperative threading strategy, where returning it yields to other threads.

Reply:
I removed `__getstate__` and `__setstate__` from `QuantumCircuit` and implemented `__reduce__` for `CircuitData` in 10cc624, which became architecturally reasonable after moving ownership of `qubits` and `clbits` from `QuantumCircuit` to `CircuitData`.

I've also removed `__copy__` (commit not yet pushed) since it is no longer needed after removing `__{get,set}state__`, which leaves us with just `__deepcopy__` added to `QuantumCircuit`.